File: 034.c
/* Digit factorials.
 * Problem 034
 *
 * 145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.
 *
 * Find the sum of all numbers which are equal to the sum of the factorial of
 * their digits.
 *
 * Note: as 1! = 1 and 2! = 2 are not sums they are not included.
 */

#include <stdio.h>

int main() {
  // NB: 0! = 1
  const int FACTORIAL[10] = { 1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880 };
  unsigned long long x = 3, n = 1, m = 9, f = 10, r, sum = 0, digit_sum;
  printf("Digit factorials:\n");
  while (n * FACTORIAL[9] > m) {
    digit_sum = FACTORIAL[x%10];
    r = x / 10;
    while (r) {
      digit_sum += FACTORIAL[r%10];
      r /= 10;
    }
    if (digit_sum == x) {
      printf("%d\n", x);
      sum += x;
    }

    ++x;
    if (x % f == 0) {
      ++n;
      m = m * 10 + 9;
      f *= 10;
    }
  }
  printf("Sum = %d\n", sum);

  return 0;
}

File: 034.py
#!/usr/bin/env python

"""
Digit factorials.
Problem 034

145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.

Find the sum of all numbers which are equal to the sum of the factorial of
their digits.

Note: as 1! = 1 and 2! = 2 are not sums they are not included.
"""

# Precalculate factorial look up table, since we need only 1! - 9!
# NB: 0! = 1
factorials = [1]
for i in range(1, 10):
  f = 1
  for j in range(i, 0, -1):
    f *= j
  factorials.append(f)

# Max from the fact that 7-digit 9999999 > 7 * 9!
fact_sum = 0
print("Digit factorials:")
for x in range(3, 2540160):
  digit_sum = sum(list(factorials[int(i)] for i in str(x)))
  if digit_sum == x:
    print(x)
    fact_sum += x
    x += 1
    f = 10
  elif digit_sum > x:
    x //= f
    x += 1
    x *= f
    f *= 10
  else:
    f = 10
    x += 1
print("Sum =", fact_sum)
